File: agent/llm_log.py
"""LangChain callback that logs every Liquid/OpenRouter chat call made through agent.llm.liquid_chat_model:
`llm_call_done` (model, purpose, prompt/response chars, input/output/reasoning tokens, latency) or `llm_call_failed`.
Prompts are not logged, only a 200-char preview of the last message."""
import logging
import sys
import threading
import time

from langchain_core.callbacks import BaseCallbackHandler

from core.logs import event

log = logging.getLogger("agent.llm")
GENERIC = {"call_llm", "_call", "ask", "ask_json", "call", "invoke", "run", "_run", "<lambda>", "result"}


def caller_purpose() -> str:
    """Name of the first agent/core function up the stack that is not a generic wrapper (e.g. `extract_page`)."""
    frame = sys._getframe(1)
    fallback = None
    while frame is not None:
        path = frame.f_code.co_filename.replace("\\", "/")
        name = frame.f_code.co_name
        if ("/agent/" in path or "/core/" in path) and not path.endswith("llm_log.py"):
            if name not in GENERIC:
                return name
            fallback = fallback or name
        frame = frame.f_back
    return fallback or "unknown"


def _text(content) -> str:
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return " ".join(p.get("text", "") if isinstance(p, dict) else str(p) for p in content)
    return str(content or "")


class LlmLogHandler(BaseCallbackHandler):
    def __init__(self, model: str):
        self.model = model
        self.calls = {}
        self.lock = threading.Lock()

    def on_chat_model_start(self, serialized, messages, *, run_id, metadata=None, **kwargs):
        flat = [m for batch in messages for m in batch]
        chars = sum(len(_text(getattr(m, "content", ""))) for m in flat)
        purpose = (metadata or {}).get("purpose") or caller_purpose()
        with self.lock:
            self.calls[run_id] = (time.perf_counter(), chars, purpose)
        event(log, "llm_call_started", logging.DEBUG, model=self.model, purpose=purpose, messages=len(flat),
              promptChars=chars, prompt=_text(getattr(flat[-1], "content", "")) if flat else "")

    def _pop(self, run_id):
        with self.lock:
            return self.calls.pop(run_id, (time.perf_counter(), 0, "unknown"))

    def on_llm_end(self, response, *, run_id, **kwargs):
        t0, chars, purpose = self._pop(run_id)
        usage, text = {}, ""
        try:
            gen = response.generations[0][0]
            message = getattr(gen, "message", None)
            text = _text(getattr(message, "content", None) or getattr(gen, "text", ""))
            usage = dict(getattr(message, "usage_metadata", None) or {})
        except (IndexError, AttributeError, TypeError):
            pass
        if not usage:
            raw = (getattr(response, "llm_output", None) or {}).get("token_usage") or {}
            usage = {"input_tokens": raw.get("prompt_tokens"), "output_tokens": raw.get("completion_tokens"),
                     "total_tokens": raw.get("total_tokens")}
        details = usage.get("output_token_details") or {}
        event(log, "llm_call_done", model=self.model, purpose=purpose, promptChars=chars, responseChars=len(text),
              inputTokens=usage.get("input_tokens"), outputTokens=usage.get("output_tokens"),
              reasoningTokens=details.get("reasoning") or None, totalTokens=usage.get("total_tokens"),
              durationMs=int((time.perf_counter() - t0) * 1000))

    def on_llm_error(self, error, *, run_id, **kwargs):
        t0, chars, purpose = self._pop(run_id)
        event(log, "llm_call_failed", logging.WARNING, model=self.model, purpose=purpose, promptChars=chars,
              error="{}: {}".format(type(error).__name__, str(error)[:300]),
              durationMs=int((time.perf_counter() - t0) * 1000))

File: generation-video/src/lib/llm-log.ts
/*
 * Logging for OpenRouter chat calls (see openrouter.ts): `llm_call_started` (debug), `llm_call_retry` (warn),
 * `llm_call_done` (info: model, purpose, prompt/response chars, tokens, latency, streamed) and `llm_call_failed`.
 * Prompts are never logged in full — only a 200-char preview of the last user message.
 */
import { log, promptPreview, startTimer } from "@/lib/runtime-log";

export type LlmUsage = { promptTokens?: number; completionTokens?: number; totalTokens?: number; cost?: number };

type Message = { role: string; content: string | ({ type: string; text?: string } | { type: string })[] };
type Result = { content: string; toolCalls: { function?: { name?: string } }[]; finishReason?: string; model: string; usage?: LlmUsage };

function textOf(content: Message["content"]) {
  if (typeof content === "string") return content;
  return content.map((part) => ("text" in part && typeof part.text === "string" ? part.text : "")).join(" ");
}

const SKIP_FRAMES = /(llm-log|openrouter|createChatCompletion|callerPurpose|logLlmCall|processTicksAndRejections|new Promise|Array\.map|Promise\.all|<anonymous>)/;

/** Best-effort name of the function that called createChatCompletion (e.g. `detectIntent`), for the `purpose` field. */
export function callerPurpose() {
  const stack = new Error().stack?.split("\n").slice(1) ?? [];
  for (const frame of stack) {
    const match = /at (?:async )?([\w$.<>]+) \(/.exec(frame.trim());
    if (!match || SKIP_FRAMES.test(frame)) continue;
    const name = match[1].replace(/^Object\./, "").replace(/^Module\./, "");
    if (name && name !== "eval" && !/^(?:__|_)/.test(name)) return name;
  }
  return "unknown";
}

export function logLlmCall(input: { model: string; purpose: string; messages: Message[]; tools?: number; streamed: boolean; maxTokens?: number }) {
  const elapsed = startTimer();
  const promptChars = input.messages.reduce((sum, message) => sum + textOf(message.content).length, 0);
  const images = input.messages.reduce((sum, message) => sum + (Array.isArray(message.content) ? message.content.filter((part) => part.type === "image_url").length : 0), 0);
  const lastUser = [...input.messages].reverse().find((message) => message.role === "user");
  const base = { model: input.model, purpose: input.purpose, streamed: input.streamed };
  let attempts = 1;
  log("debug", "llm_call_started", { ...base, messages: input.messages.length, promptChars, images: images || undefined, tools: input.tools, maxTokens: input.maxTokens, prompt: promptPreview(lastUser ? textOf(lastUser.content) : "") });
  return {
    attempt(index: number, status: number) {
      attempts = index + 1;
      if (status >= 400) log(index < 2 && [429, 502, 503].includes(status) ? "warn" : "debug", "llm_call_retry", { ...base, attempt: attempts, status, afterMs: elapsed() });
    },
    memoized(result: Result) {
      log("info", "llm_call_done", { ...base, memoized: true, responseChars: result.content.length, durationMs: elapsed() });
    },
    done(result: Result) {
      log("info", "llm_call_done", {
        ...base,
        responseModel: result.model !== input.model ? result.model : undefined,
        promptChars,
        responseChars: result.content.length,
        toolCalls: result.toolCalls.length ? result.toolCalls.map((call) => call.function?.name ?? "?").join(",") : undefined,
        finishReason: result.finishReason,
        promptTokens: result.usage?.promptTokens,
        completionTokens: result.usage?.completionTokens,
        totalTokens: result.usage?.totalTokens,
        cost: result.usage?.cost,
        attempts: attempts > 1 ? attempts : undefined,
        durationMs: elapsed(),
      });
    },
    failed(error: unknown) {
      const aborted = error instanceof Error && error.constructor?.name === "ClientAbortedError";
      log(aborted ? "info" : "warn", aborted ? "llm_call_aborted" : "llm_call_failed", {
        ...base,
        attempts,
        status: (error as { status?: unknown })?.status as number | undefined,
        error: error instanceof Error ? error.message : String(error),
        durationMs: elapsed(),
      });
    },
  };
}
